### PR TITLE
[XLA] Change the number of threads per block for row reduction

### PR DESCRIPTION
**Pull request by nouiz**
This is a follow up from https://github.com/tensorflow/tensorflow/pull/36752

This change the number of blocks per threads. It is only the last commit that is the new feature from this PR. The others are already in 36752.

I didn't change anything in the code. We can start the discussion on how to fix this and I'll update the code accordingly.

The number of threads per blocks must be determined by many factors (taken from the previous PR)

- More threads per blocks mean less atomic writes (as the total number of threads is constant). This is good.
- One SM can't use the full bandwidth per itself. So you need enough blocks to help saturate the IO. I did very small timing, but discarded those results as the noise was too high. There is 80 SM in V100, so ideally we would like at least 80 blocks.
- The last block strangler effect. If you have 81 blocks and 80 SM, then you have 1 that will cause the computation to be as long as if you had 160 blocks. More blocks mean smaller blocks, help mitigate this, as the smaller block will run faster.
- Also, we need at least 64 threads per blocks if we want full occupancy. We are limited to 0.5 occupancy if we use 32 threads per blocks.
- In theory, we could also add a cost model to know if we have a IO or Compute bound kernel (fusion could create compute bound kernel). And in that case we could need different best configuration, as it could be register bound... I do not want to go there.

The problem this PR currently have is that it breaks the deterministic reduction on XLA as it is expect exactly 1024 threads per blocks. Making the tree reduction opt know about the new block size is easy (already tested). But this still cause 2 tree reduction failure as now it request 1 extra reduction. This probably slowing down the current tree reduction.

- Is slowing down the tree reduction acceptable for this PR? If so, I'll just update the expected answer from those tests.
- One option would be to add an attribute to the XLA reduction that would specify the number of threads (example name biggest_block).
    - This isn't very good as this add an implementation detail as a Reduce attribute.
- Another attribute that we could add is: "deterministic". This describe a behavior. I could implement that, by making sure we it generates only 1 blocks per row. If not possible, raise an error. It would be the responsibility of the optimizer to have this possible.
  
I like the last approch. It would also help prevent the AlgebraicOptimizer to interfer with the TreeReductionRewriter optimizer, as it could just not touch those graph.

If we choose this options, this open the doors to more optimization later. For example, you could remove the memset and replace the atomic update with a normal write at the end of the kernel.

@cheshire

### CODE REVIEW

**Comment by nouiz** [comment]
I just pushed a new commit that make the tree reduction use the right information following the changes to XLA reduction. Now it will always use the same unroll factor and block size.
So you should be able to start the benchmark from this commit.

The reduction tests pass, but I have a strange error/warning from bazel tests. I do not understand the error information. Any idea what could be going on?
```
ERROR: command succeeded, but there were errors parsing the target pattern
INFO: Elapsed time: 25.840s, Critical Path: 18.46s
INFO: 39 processes: 39 local.
FAILED: Build did NOT complete successfully
//tensorflow/compiler/xla/service/gpu/tests:gpu_unrolling_test           PASSED in 5.4s
//tensorflow/compiler/xla/service/gpu/tests:reduction_degenerate_dim_remover_test PASSED in 5.0s
//tensorflow/compiler/xla/service/gpu/tests:reduction_dimension_grouper_test PASSED in 4.9s
//tensorflow/compiler/xla/service/gpu/tests:reduction_layout_normalizer_test PASSED in 5.4s
//tensorflow/compiler/xla/service/gpu/tests:tree_reduction_rewriter_test PASSED in 17.7s
//tensorflow/compiler/xla/tests:reduce_hlo_test_gpu                      PASSED in 5.5s
//tensorflow/compiler/xla/tests:reduce_precision_test_gpu                PASSED in 4.9s
//tensorflow/compiler/xla/tests:vector_ops_reduce_test_gpu               PASSED in 4.8s
//tensorflow/compiler/xla/tests:reduce_test_gpu                          PASSED in 5.5s
  Stats over 31 runs: max = 5.5s, min = 2.1s, avg = 4.2s, dev = 0.9s

Executed 9 out of 9 tests: 9 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these
are.
All tests passed but there were other errors during the build.
FAILED: Build did NOT complete successfully
```

**Comment by nouiz** [comment]
I rebased this PR to resolve the conflicts.
Note, I didn't update the tests, this PR is in a good state to be benchmarked, but not merged.

File: tensorflow/compiler/xla/service/gpu/tree_reduction_rewriter.cc [hunk 1]
```diff
@@ -97,9 +97,27 @@ class ReductionRewriterVisitor : public DfsHloRewriteVisitor {
                                                  input_shape, input);
     }
     bool is_row_reduction = reduction_dimensions.is_row_reduction;
-
-    int64 atomic_free_bound =
-        is_row_reduction ? kRowAtomicFreeBound : kColumnAtomicFreeBound;
+    int64 atomic_free_bound = 0;
+    if (is_row_reduction) {
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/tree_reduction_rewriter.cc]: Please add a comment to explain how the atomic free bound is determined for row reductions.

File: tensorflow/compiler/xla/service/gpu/tests/tree_reduction_rewriter_test.cc [hunk 2]
```diff
@@ -70,7 +70,7 @@ ENTRY main {
 )";
 
   // TODO(cheshire): a more generic check, do not hardcode the names.
-  MatchOptimizedHloWithShapes(hlo_text,
+  /*  MatchOptimizedHloWithShapes(hlo_text,
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/tests/tree_reduction_rewriter_test.cc]: You probably missed that this test and another one below are still commented out.

**Review by thomasjoerg — DISMISSED** [review]
Minor comment. @cheshire it'd be great if you could take a look too, since you are most familiar with the code being changed.

File: tensorflow/compiler/xla/service/gpu/tests/tree_reduction_rewriter_test.cc (hunk 2 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/tests/tree_reduction_rewriter_test.cc]: This PR isn't ready to merge, just ready to benchmarks. As it isn't sure we will merge it as is, I didn't fix the tests. We need to figure out if this PR hurts the tree reductions. If so, I can make a slightly more complication version that won't hurt it. But @cheshire isn't sure the complication is useful. So he wanted to benchmark it first.

Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/tests/tree_reduction_rewriter_test.cc]: Done. Maybe the interface can be simplifier, but before doing it, I would like to know if I finish it like this or I make it a little bit more complicated to not slowdown the tree reduction.

**Comment by thomasjoerg** [comment]
Talked to @cheshire who has the most context here and agreed to take a look. I'm going to remove myself as reviewer.

**Comment by gbaned** [comment]
@nouiz Can you please resolve conflicts? Thanks!

**Comment by nouiz** [comment]
I have rebased it.

**Comment by gbaned** [comment]
@cheshire Can you please review this PR ? Thanks!

**Comment by cheshire** [comment]
@nouiz I'm generally fine with merging this. Could you add some micro-benchmarks in the commit description demonstrating a speed-up in small artificial test cases?

**Comment by nouiz** [comment]
I'll work on this probably at the end of this week.

**Comment by gbaned** [comment]
@nouiz  Any update on this PR? Please. Thanks!

**Comment by nouiz** [comment]
I'm not able to repro it anymore.  I'm not sure what is going on. As the gain was only for small case, this isn't the most important and I do not have time now to investigate this.
So I'll close this PR and if I get back to it, I'll make a new one or reopen it.

**Pull request closed without merging**